PDF Generation

aivantg opened the pull request:
[//]: # "These comments meant for your reference, they are invisible and don't need to be deleted"

## What's new in this PR?

Added PDF Generation to the Bill Generation Process. 

Process: 
1. After creating bill, generate PDF using react-renderer
2. Save to local file (which is openly accessible on the internet)
3. Upload to Airtable using public link
4. After 20 seconds, delete the file (to ensure airtable had time to copy it over

### How to Review

Check out the code! 

To test, delete existing subscriber bill records and log in to the frontend using 'grayson@gmail.com' and generate bills for Lora Jo Foo's house. You should see the PDF Generate and get attached to the airtable record! 

### Related PRs

[//]: # "Optional - related PRs you're waiting on/ PRs that will conflict, etc"

### Next steps

- Calculations...might be off. need to compare with google sheet because the numbers we're getting are enormous
- Might need to look into [this article](https://medium.com/@kimtnguyen/how-to-deploy-es6-node-js-express-back-end-to-heroku-7e6743e8d2ff) to make our deployment less janky because we are currently using `babel-node` which is not recommended in production. But I'm not stressed because our use case is relatively small tbh

### Tests Performed, Edge Cases

[//]: # 'If you made changes/additions to `airtable.js` make sure you update/add'
[//]: # '[at least 1] corresponding unit tests in `airtable.spec.js`'

### Screenshots

[//]: # "Add screenshots!!! If you'd like"

CC: @dfangshuo

dfangshuo commented on index.js:
```diff
@@ -10,6 +10,7 @@ const port = process.env.PORT || 3000;
 
 app.use(cors());
 app.use(express.json());
+app.use(express.static('temp')); // Make PDFs accessible
```
all pdfs would be saved to a directory called `temp` then, maybe we could name this something more descriptive?

aivantg replied:
well it's not a final save, it's a temporary save. it deletes it after 30 min. i think temp is a good name because in the future (if we generate charts) we can save other temp files here too! 

My thought is, if we end up needing to expose other static files, we can make a better name, but for now this suffices

dfangshuo replied:
agree with that!

dfangshuo commented on utils/billgeneration.js:
```diff
@@ -1,20 +1,68 @@
+import BillingTemplate from './pdf/BillingTemplate';
+
+// const ROOT_URL = 'https://1f7c24d2.ngrok.io';
+const ROOT_URL = 'https://peoplepower-node.herokuapp.com';
```
I poked around the codebase a little and noticed that all in all we have quite a number of constants decalred in various js files. Given the way the codebase has grown perhaps it might be a good idea to now add all these constants into a `constants.js` file or something. I'd put this as a nit though

aivantg replied:
yeah maybe. will look into

dfangshuo commented on utils/billgeneration.js:
```diff
@@ -1,20 +1,68 @@
 import getLatestBill from './utilityApi';
+import BillingTemplate from './pdf/BillingTemplate';
+
+// const ROOT_URL = 'https://1f7c24d2.ngrok.io';
```
what was this? I'd remove it if it's nthing too significant

aivantg replied:
it's a local testing url, but yup sure good point

dfangshuo commented on utils/pdf/BillingTemplate.js:
```diff
@@ -0,0 +1,383 @@
+    backgroundColor: 'white'
+  },
+  gray10: {
+    color: '#747474'
```
i think it's a good idea to start a `main.css` or `colors.css` file now to avoid having to deal with consistency issues down the road. Ditto for all hex codes declared below

aivantg replied:
i guesssssss so :P 

Will have to ask Ashley why all these styles aren't actually in a css file speaking of which

dfangshuo replied:
ahh frk if it's in a css file i'd rather it be in kebab case. but it's fine, we can address that in a future PR

dfangshuo commented on utils/pdf/BillingTemplate.js:
```diff
@@ -0,0 +1,383 @@
+            <View>
+              <Image
+                src="./assets/PPSC-logo-no-padding.png"
+                allowDangerousPaths
```
this might be a question for @ashleynguyen27 can we explain the motivation for using this instead of defining safePaths?

aivantg replied:
this was actually me, whoops lol I'll see if i can find a "safe" way to do it, i was just trying to get it to work.

dfangshuo commented on utils/pdf/BillingTemplate.js:
```diff
@@ -0,0 +1,383 @@
+            <View style={[styles.flex, styles.justifyContent]}>
+              <View style={[styles.left]}>
+                <Text style={[styles.text]}>1234 Address St.</Text>
+                <Text style={[styles.text]}>City, CA 12345</Text>
```
these values are still hardcoded? :o

aivantg replied:
I believe this will eventually be the address of peoplepower......but i'll look at the original bill and make sure

dfangshuo commented on utils/billgeneration.js:
```diff
@@ -1,20 +1,68 @@
+  setTimeout(() => {
+    console.log(`Deleting Temporary PDF: ${subscriberBill.id}.pdf`);
+    fs.unlinkSync(`./temp/${subscriberBill.id}.pdf`);
+  }, 20000);
```
i guess because we're using async await here, we will only ever cal this line after we've successfully updated the subscriber bill? Do we need the timeout then in this case?

aivantg replied:
Yeah so what I learned was that, when you upload a url to airtable, it downloads the pdf and then reuploads it to its own servers. 

So the PDF needs to exist on our servers for long enough for airtable to copy it over. I arbitrarily set it at 20 seconds before it deletes.

dfangshuo replied:
wait that's so clean. wtf airtable ++ . I guess a potential issue here then, is that there might be cases when airtable doesn't actually download it in 20s, but that's an edge case we can address in the future. 20000 in Constants.js so we can change it easily in the future?

aivantg replied:
sounds good

aivantg replied:
I think if it it doesn't do it in 20 seconds you can just regenerate but moved to constants

dfangshuo commented on utils/billgeneration.js:
```diff
@@ -1,20 +1,68 @@
+  setTimeout(() => {
+    console.log(`Deleting Temporary PDF: ${subscriberBill.id}.pdf`);
+    fs.unlinkSync(`./temp/${subscriberBill.id}.pdf`);
+  }, 20000);
```
I wanna make sure I fully understand what's going on here: 

ReactPDF.render automatically saves a file to storage (which is in herokue, in this case). we then update the subscriber bill with a unique url? Where in the code is that pdf actually 'uploaded' to that url? 

After successfullyl doing so, we delete that pdf

aivantg replied:
great question! 

so `ReactPDF.render` is saving it in the local folder called `temp`. In `index.js`, the line `app.use(express.static('temp'))` _exposes_ that `temp` folder to the web. Aka any files in that `temp` folder will be available at `peoplepower-node.herokuapp.com/filename`. Then, we give that url to airtable, it copies the pdf over, and we delete it once it's done

aivantg commented:
@dfangshuo I added the updates. Things fixed: 
- styles moved to their own folder, i didn't move colors to variables because realistically they weren't repeated a lot, but i did remove unnecessary classes
- i fixed the due date problem
- i moved some things to constants but....there was actually only one constant in the entire app LOL
